### ISSUE TITLE
Add a tree table view widget

### ISSUE THREAD

**Issue by ddkwork** [cogentcore/core#675]
![{F93A9AAD-6848-4b81-9DAB-566F38C17683}.png](https://github.com/goki/gi/assets/19886504/8abe012d-257d-4f2f-9b07-8d82d2dd83c5)

![{20139C18-563C-4da9-9061-5B84559A7ADC}.png](https://github.com/goki/gi/assets/19886504/75e6a283-ebb5-4276-987c-563552c9b843)

**Comment by ddkwork** [comment]
gcs Repositories is here :
https://githubfast.com/richardwilkes/gcs

**Comment by ddkwork** [comment]
![Screenshot_20231228_121841_com.realvnc.viewer.android.jpg](https://github.com/goki/gi/assets/19886504/5e34fece-4f53-41cd-b981-683b83129c21)

![1703737293530.png](https://github.com/goki/gi/assets/19886504/efc71424-0b26-443e-9d8d-b1a3ee931b82)

![1703737307160.png](https://github.com/goki/gi/assets/19886504/c341ab02-7ec7-497f-b4cb-38b52de8f3d3)

More and more scenarios require tree tables

**Comment by kkoreilly** [comment]
This is on our list of things to do, but we have to finish other things first. It should be done relatively soon though.

**Comment by ddkwork** [comment]
It's time to implement these widgets, I'd love to see how you draw this table step by step, I'm going to go through every single line of your code, at the moment almost all the panic is controllable, stable, please implement it if you have time, so I can learn your drawing principles, I know I can't help much at the moment, it's bad, once I get the drawing principles, there's a lot of issues left in the list I think I can get involved in working on it

**Comment by kkoreilly** [comment]
I will work on implementing the tree table at some point relatively soon after I finish some other things.

**Comment by ddkwork** [comment]
After it is implemented, I think we can redesign the appearance of the files app
<img width="1276" alt="{01693CFF-71B3-40f7-BD60-2DF19F126F9E}" src="https://github.com/cogentcore/core/assets/19886504/7c258260-ba1d-4287-b02d-0cedc4400da1">

**Comment by ddkwork** [comment]
```go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/ddkwork/golibrary/stream"
)

type Data struct {
	name string
	size int64
	typ  os.FileMode
	time string
}

type Node struct {
	Data     Data
	parent   *Node
	children []*Node
}

func toNodeSlice(path string, parent *Node) ([]*Node, error) {
	entries, err := os.ReadDir(path)
	if err != nil {
		return nil, err
	}
	nodes := make([]*Node, len(entries))
	for i, entry := range entries {
		info, err := entry.Info()
		if err != nil {
			fmt.Println(err)
			continue
		}
		data := Data{
			name: entry.Name(),
			size: info.Size(),
			typ:  entry.Type(),
			time: stream.FormatTime(info.ModTime()),
		}

		node := &Node{Data: data, parent: parent}
		if entry.IsDir() {
			child, err := toNodeSlice(filepath.Join(path, entry.Name()), node)
			if err != nil {
				fmt.Println(err)
				continue
			}
			node.children = child
		}
		nodes[i] = node
	}
	return nodes, nil
}

func PrintTree(root *Node) {
	printTree(root, "", true)
}

func printTree(root *Node, prefix string, isLast bool) {
	fmt.Printf("%s", prefix)
	if isLast {
		fmt.Printf("└───")
		prefix += "    "
	} else {
		fmt.Printf("├───")
		prefix += "│   "
	}
	fmt.Printf("%s. %d (%s): %v\n",
		root.Data.name,
		root.Data.size,
		root.Data.typ.String(),
		root.Data.time,
	)

	for i := 0; i < len(root.children); i++ {
		printTree(root.children[i], prefix, i == len(root.children)-1)
	}
}

func main() {
	nodes, err := toNodeSlice(".", nil)
	if err != nil {
		fmt.Println(err)
		return
	}
	for _, node := range nodes {
		PrintTree(node)
	}
	// printNodeSlice(nodes, 0)
	fmt.Println("=====")
	fmt.Println(nodes[0].parent)
	// fmt.Println(nodes[0].children[0].parent)
}

```

**Comment by ddkwork** [comment]
Then the process becomes like this: pass an n-fork tree struct pointer to the treetableView widget, run configWidhet at this time, and get the labels and values of each parent and child fields by reflecting the struct, which is to put it bluntly, all the fields of the nested struct. In this case, if it is a container node, take the coordinates of the body minus the left and right sides to fill the row width to the container node, and if it is a non-container node, start on the left side + deep to calculate the row width of the non-container node. Finally, set the height to a value, and perform depth traversal and breadth traversal on the n-fork tree reflected by the structure, and finally calculate the width, height and coordinates of all rows, anyway, the up and down are connected. Once you have the coordinates and size of all the rows, you can draw a rectangle? Stuffing mouse and key events, and changing the dynamic insertion of rows triggered by structs via context menu events? I don't know if it's such a process, it's 🙄 difficult, and there are header sorting, column operations, etc., it's too difficult

**Comment by kkoreilly** [comment]
Please do not worry. I have a plan for how to implement the tree table, and I will work on it as soon as I have time. The tree table is a lower priority for me than other things that have to happen before a v1 release, but I should get around to it in not too long.

**Comment by ddkwork** [comment]
okay



---Original---
From: ***@***.***&gt;
Date: Thu, Mar 7, 2024 13:14 PM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"State ***@***.***&gt;;
Subject: Re: [cogentcore/core] Add a tree table view widget (Issue #675)




 
Please do not worry. I have a plan for how to implement the tree table, and I will work on it as soon as I have time. The tree table is a lower priority for me than other things that have to happen before a v1 release, but I should get around to it in not too long.

**Comment by ddkwork** [comment]
```go
 /*
  ──────────────────────────────────────────────────────────────────────────────────────────────────
  │Method                                                                          │Description
  ──────────────────────────────────────────────────────────────────────────────────────────────────
  │doc_container                                                                   │
  ├───init_container                                                               │
  │   ├───New[T any](typeKey string, data T) (root *Node[T])                       │
  │   ├───NewContainerNode[T any](typeKey string, data T) *Node[T]                 │
  │   ├───NewNode[T any](typeKey string, data T) *Node[T]                          │
  │   ├───Root() *Node[T]                                                          │
  │   └───IsRoot() bool	                                                           │
  ├───walker_container                                                             │
  │   ├───Walk(callback func(node *Node[T]))                                       │
  │   └───WalkContainer(callback func(node *Node[T]))                              │
  ├───ctx_container                                                                │
  │   ├───Depth() int                                                              │
  │   ├───UUID() uuid.UUID                                                         │
  │   ├───Sort(cmp func(a T, b T) bool)                                            │
  │   ├───Clone(newParent *Node[T], preserveID bool) *Node[T]                      │
  │   ├───CopyFrom(from *Node[T])                                                  │
  │   ├───ApplyTo(to *Node[T])                                                     │
  │   ├───Parent() *Node[T]                                                        │
  │   ├───SetParent(parent *Node[T])                                               │
  │   ├───Container() bool                                                         │
  │   ├───HasChildren() bool                                                       │
  │   ├───SetChildren(children []*Node[T])                                         │
  │   ├───clearUnusedFields()                                                      │
  │   ├───GetType() string                                                         │
  │   ├───SetType(t string)                                                        │
  │   ├───kind(base string) string                                                 │
  │   ├───Open() bool                                                              │
  │   ├───SetOpen(open bool)                                                       │
  │   ├───OpenAll() bool                                                           │
  │   ├───CloseAll() bool                                                          │
  │   ├───LenChildren() int                                                        │
  │   ├───LastChild()                                                              │
  │   ├───IsLastChild() bool                                                       │
  │   └───ResetChildren()                                                          │
  ├───adder_container                                                              │
  │   ├───AddContainer(container *Node[T])                                         │
  │   ├───AddContainerByData(typeKey string, data T)                               │
  │   ├───AddChildByData(typeKey string, data T)                                   │
  │   ├───AddChild(child *Node[T])                                                 │
  │   ├───InsertChildItem(parentID uuid.UUID, data T) *Node[T]                     │
  │   ├───CreateItem(parent *Node[T], data T)                                      │
  │   ├───InsertChildByID(id uuid.UUID) *Node[T]                                   │
  │   ├───InsertChildByName(id uuid.UUID) *Node[T]                                 │
  │   ├───InsertChildByIndex(id uuid.UUID) *Node[T]                                │
  │   ├───InsertChildByType(id uuid.UUID) *Node[T]                                 │
  │   └───InsertChildByPath(id uuid.UUID) *Node[T]                                 │
  ├───remover_container                                                            │
  │   ├───RemoveByID(id uuid.UUID)                                                 │
  │   ├───RemoveSelf(id uuid.UUID)                                                 │
  │   ├───RemoveChildByID()                                                        │
  │   ├───RemoveChildByName()                                                      │
  │   ├───RemoveChildByIndex()                                                     │
  │   └───RemoveChildByPath()                                                      │
  ├───updater_container                                                            │
  │   ├───UpdateChildByID(id uuid.UUID, data T)                                    │
  │   ├───UpdateChildByName(id uuid.UUID, data T)                                  │
  │   ├───UpdateChildByIndex(id uuid.UUID, data T)                                 │
  │   ├───UpdateChildByPath(id uuid.UUID, data T)                                  │
  │   ├───UpdateChildByType(id uuid.UUID, data T)                                  │
  │   ├───SwapChild()                                                              │
  │   ├───MoveChildTo()                                                            │
  │   └───SwapChild()                                                              │
  ├───finder_container                                                             │
  │   ├───FindChildByID(id uuid.UUID) *Node[T]                                     │
  │   ├───FindChildByName(id uuid.UUID) *Node[T]                                   │
  │   ├───FindChildByIndex(id uuid.UUID) *Node[T]                                  │
  │   ├───FindChildByPath(id uuid.UUID) *Node[T]                                   │
  │   └───FindChildByType(id uuid.UUID) *Node[T]                                   │
  └───tableCtx_container                                                           │
  │   ├───SetRowCallback(rowCallback func(node *Node[T]) (cell []string))           │
  │   ├───String() string                                                           │
  │   ├───Document() string                                                         │
  │   ├───format(node *Node[T], s *stream.Stream)                                   │
  │   ├───Enabled() bool                                                            │
  │   ├───SetSelectedFunc()                                                         │
  │   ├───SetIndent()                                                               │
  │   ├───GetPath()                                                                 │
  │   ├───SetAlign()                                                                │
  │   └───MouseHandler()                                                            │
*/

/*
  ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  │Number                         │Name               │Type │Kind    │Tag │Value │ValueAssert
  ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  │1                              │Group1_container   │     │struct  │    │      │<nil>
  ├───1                           │Binary1            │     │string  │    │      │game/system/session/info
  ├───2                           │Message2_container │     │struct  │    │      │<nil>
  │   ├───2                       │Packed2_container  │     │slice   │    │      │<nil>
  │   │   ├───2                   │[0]_container      │     │invalid │    │      │<nil>
  │   │   │   ├───1               │Varint1            │     │uint64  │    │      │0
  │   │   │   ├───2               │Binary2            │     │string  │    │      │d3048a459417e6c0b7d39c971b99a58029f2720f7b2a70c992c826ce48184069
  │   │   │   ├───3               │Binary3            │     │string  │    │      │6593D03B-92BC-4BC6-BF54-D16BB0271AF9
  │   │   │   └───4               │Binary4            │     │string  │    │      │Apple-iPhone10,3
  │   │   └───2                   │[1]_container      │     │invalid │    │      │<nil>
  │   │   │   ├───1               │Varint1            │     │uint64  │    │      │1
  │   │   │   ├───2               │Binary2            │     │string  │    │      │""
  │   │   │   ├───3               │Binary3            │     │string  │    │      │EAAIe5YPC68wBAPGE5l7JEtIE9BfPQmbcpQ92b0c9fD29vKn5ZCwHkutEjpX2PEcvyBLDqo15gNi1x0VN7U6d26QDaABEDaVzu3vuZBYKtvHH130O9Kna4742s6B8dtr1aKJUw7HuuyNWObpWYZCqBDGvypB7Js93oBISkwWrjbuYZAooY5vHLHFPyuIcLAcV8ZAX9sRFn3un18KZA0MEuDtgekU4ZBTJ1doEQmQ4DCLXwZDZD
  │   │   │   └───4               │Binary4            │     │string  │    │      │_
  │   └───3                       │Binary3            │     │string  │    │      │6593D03B-92BC-4BC6-BF54-D16BB0271AF9
  ├───3                           │Varint3            │     │uint64  │    │      │0
  ├───4                           │Binary4            │     │string  │    │      │""
  └───5                           │Varint5            │     │uint64  │    │      │0
*/
```

**Comment by ddkwork** [comment]
I seem to be getting the correct layout and algorithm for calculating column widths, looking forward to your updates!

**Comment by ddkwork** [comment]
```go
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
│Name                                          │Size      │Hash         │UpdateTime           │Description                                                                                                                                                                              
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
│Models_container                              │2392.50GB │             │                     │                                                                                                                                                                                         
├───gemma_container                            │290.90GB  │             │                     │Gemma is a family of lightweight, state-of-the-art open models built by Google DeepMind.                                                                                                 
│   ├───gemma:latest                           │5.2GB     │430ed3535049 │Updated 5 days ago   │                                                                                                                                                                                         
│   ├───gemma:2b                               │1.7GB     │b50d6c999e59 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b                               │5.2GB     │430ed3535049 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:instruct                         │5.2GB     │430ed3535049 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:text                             │5.2GB     │af57093b878e │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct                      │1.7GB     │b50d6c999e59 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text                          │1.7GB     │4003359bdf67 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct                      │5.2GB     │430ed3535049 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text                          │5.2GB     │af57093b878e │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q4_0                 │1.7GB     │b50d6c999e59 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q4_1                 │1.8GB     │79f47baf629d │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q5_0                 │1.9GB     │f2e0978e20a7 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q5_1                 │2.1GB     │2005cd5b9f9a │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q8_0                 │2.7GB     │233b6c11a9a9 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q2_K                 │1.3GB     │462aa639add9 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q3_K_S               │1.4GB     │1147ae8910c1 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q3_K_M               │1.5GB     │0bb5a6a99256 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q3_K_L               │1.6GB     │a5e1a96f2947 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q4_K_S               │1.7GB     │1f62e695daaf │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q4_K_M               │1.8GB     │073d0876d678 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q5_K_S               │1.9GB     │466d55f10b72 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q5_K_M               │2.0GB     │79fe7a4a6e26 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-q6_K                 │2.2GB     │2934d89f4873 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-instruct-fp16                 │4.5GB     │1cca4a1914b8 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q4_0                     │1.7GB     │4003359bdf67 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q4_1                     │1.8GB     │2336b3481fc2 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q5_0                     │1.9GB     │b82a85f35c76 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q5_1                     │2.1GB     │af9d160f687e │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q8_0                     │2.7GB     │40e5082de8d9 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q2_K                     │1.3GB     │fa08f447fb74 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q3_K_S                   │1.4GB     │30edc7975a67 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q3_K_M                   │1.5GB     │3755aa666a62 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q3_K_L                   │1.6GB     │fb3cb86ed3a0 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q4_K_S                   │1.7GB     │e1f170c66fed │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q4_K_M                   │1.8GB     │14b9181dd33c │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q5_K_S                   │1.9GB     │dbbce6ea20a0 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q5_K_M                   │2.0GB     │75095004592a │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-q6_K                     │2.2GB     │5bd4a449bdbe │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:2b-text-fp16                     │4.5GB     │bced1b198253 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q4_0                 │5.2GB     │430ed3535049 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q4_1                 │5.7GB     │e1e4a2e0c8ee │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q5_0                 │6.2GB     │de923fde2f26 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q5_1                 │6.7GB     │2e51588baf45 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q8_0                 │9.1GB     │61d0f0df3637 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q2_K                 │3.7GB     │831e95226882 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q3_K_S               │4.2GB     │e4aea70c287e │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q3_K_M               │4.6GB     │92dd270cb673 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q3_K_L               │4.9GB     │1fa1f8e1a003 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q4_K_S               │5.2GB     │bb7b8325814d │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q4_K_M               │5.5GB     │d9c26a968eb4 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q5_K_S               │6.2GB     │42630cbe71a2 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q5_K_M               │6.3GB     │556abeb39bfd │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-q6_K                 │7.2GB     │5c7aded0b8bd │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-instruct-fp16                 │17GB      │f689ad351c8d │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q4_0                     │5.2GB     │af57093b878e │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q4_1                     │5.7GB     │a7adb7322d8b │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q5_0                     │6.2GB     │fd6dc849a9f8 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q5_1                     │6.7GB     │938308342124 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q8_0                     │9.1GB     │1b7b1e5e2f98 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q2_K                     │3.7GB     │e43e18a484ed │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q3_K_S                   │4.2GB     │f40e21d459ff │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q3_K_M                   │4.6GB     │ffd8ea66cf1f │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q3_K_L                   │4.9GB     │1676d1f86165 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q4_K_S                   │5.2GB     │51854f1241b3 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q4_K_M                   │5.5GB     │5ec372e05241 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q5_K_S                   │6.2GB     │df449b30e2e1 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q5_K_M                   │6.3GB     │dc8f812e49c7 │Updated 2 weeks ago  │                                                                                                                                                                                         
│   ├───gemma:7b-text-q6_K                     │7.2GB     │9e7cfd3fab5b │Updated 2 weeks ago  │                                                                                                                                                                                         
│   └───gemma:7b-text-fp16                     │16GB      │1f1e9df10872 │Updated 2 weeks ago  │                                                                                                                                                                                         
├───llama2_container                           │2101.60GB │             │                     │Llama 2 is a collection of foundation language models ranging from 7B to 70B parameters.                                                                                                 
│   ├───llama2:latest                          │3.8GB     │78e26419b446 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:text                            │3.8GB     │53b394a404fd │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b                              │3.8GB     │78e26419b446 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b                             │39GB      │e7f6c06ffef4 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:chat                            │3.8GB     │78e26419b446 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b                             │7.4GB     │d475bf4c50bc │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text                         │3.8GB     │53b394a404fd │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat                         │3.8GB     │78e26419b446 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat                        │7.4GB     │d475bf4c50bc │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text                        │7.4GB     │d9ecf5544188 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat                        │39GB      │e7f6c06ffef4 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text                        │39GB      │b9b52e6b61c5 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q4_0                    │3.8GB     │78e26419b446 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q4_1                    │4.2GB     │85eb4bc10598 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q5_0                    │4.7GB     │68cf22a36a37 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q5_1                    │5.1GB     │ae646aad1ac9 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q8_0                    │7.2GB     │fc35dd6c8370 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q2_K                    │2.8GB     │56baf05e3e67 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q3_K_S                  │2.9GB     │3af967e4a825 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q3_K_M                  │3.3GB     │63ebc93e095e │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q3_K_L                  │3.6GB     │c04042faafd0 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q4_K_S                  │3.9GB     │bec61ac554e8 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q4_K_M                  │4.1GB     │fe3834f64df5 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q5_K_S                  │4.7GB     │4f771b179264 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q5_K_M                  │4.8GB     │5918a2e3bb37 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-q6_K                    │5.5GB     │7eabddae7b75 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-chat-fp16                    │13GB      │50a12b7ddb9d │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q4_0                    │3.8GB     │53b394a404fd │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q4_1                    │4.2GB     │189a0d0904e4 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q5_0                    │4.7GB     │7377932417e0 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q5_1                    │5.1GB     │1a5e521c1e5c │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q8_0                    │7.2GB     │fd701415200a │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q2_K                    │2.8GB     │c581571a4e90 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q3_K_S                  │2.9GB     │2f3ee65fe92a │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q3_K_M                  │3.3GB     │c1ddf19ff6db │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q3_K_L                  │3.6GB     │df57d4d53dac │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q4_K_S                  │3.9GB     │1c2571d89f3b │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q4_K_M                  │4.1GB     │b50304b988e8 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q5_K_S                  │4.7GB     │6dd64259f527 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q5_K_M                  │4.8GB     │eaa320e04479 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-q6_K                    │5.5GB     │7e54061a4751 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:7b-text-fp16                    │13GB      │f69718db483a │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q4_0                   │7.4GB     │d475bf4c50bc │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q4_1                   │8.2GB     │8da45bab53a7 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q5_0                   │9.0GB     │2ceaa73b479a │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q5_1                   │9.8GB     │76c91147b201 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q8_0                   │14GB      │303a83449a06 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q2_K                   │5.4GB     │14d3396d4e1b │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q3_K_S                 │5.7GB     │224005fe00d3 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q3_K_M                 │6.3GB     │7a39765cf206 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q3_K_L                 │6.9GB     │d0a4d669c0b5 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q4_K_S                 │7.4GB     │c9bc1759c3b8 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q4_K_M                 │7.9GB     │87727269f62c │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q5_K_S                 │9.0GB     │25324e6a6e04 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q5_K_M                 │9.2GB     │34471ad29e7c │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-q6_K                   │11GB      │df2e3c78030b │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-chat-fp16                   │26GB      │83b0fe8fee34 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q4_0                   │7.4GB     │d9ecf5544188 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q4_1                   │8.2GB     │ff9b7e23cab5 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q5_0                   │9.0GB     │8f3886efc72d │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q5_1                   │9.8GB     │2afbee63a3f1 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q8_0                   │14GB      │604d6e23f647 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q2_K                   │5.4GB     │64ace2ba0375 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q3_K_S                 │5.7GB     │959925e1415d │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q3_K_M                 │6.3GB     │ff280395c1b4 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q3_K_L                 │6.9GB     │cc161efc495b │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q4_K_S                 │7.4GB     │2fadc16207a6 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q4_K_M                 │7.9GB     │74b8a493a98e │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q5_K_S                 │9.0GB     │1025495b5ee3 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q5_K_M                 │9.2GB     │4be0a0bc5acb │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-q6_K                   │11GB      │376544bcd2db │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:13b-text-fp16                   │26GB      │d19bbe682374 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q4_0                   │39GB      │e7f6c06ffef4 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q4_1                   │43GB      │8618469361ea │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q5_0                   │47GB      │6cb5828911b8 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q5_1                   │52GB      │9d1fbb3c4074 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q8_0                   │73GB      │d91d0643f080 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q2_K                   │29GB      │3cac2828f2b8 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q3_K_S                 │30GB      │051b27c23c3f │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q3_K_M                 │33GB      │f11f61dc9832 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q3_K_L                 │36GB      │e2275953b0b3 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q4_K_S                 │39GB      │e202e9dd13cc │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q4_K_M                 │41GB      │8f88029f07f8 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q5_K_S                 │47GB      │eb071689d8e5 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q5_K_M                 │49GB      │2e12d2211dd5 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-q6_K                   │57GB      │b8bbc9a59804 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-chat-fp16                   │138GB     │18d69cd1de10 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q4_0                   │39GB      │b9b52e6b61c5 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q4_1                   │43GB      │92c3da1574b4 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q5_0                   │47GB      │4ec9f6c43f82 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q5_1                   │52GB      │9cafaca2d2fe │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q8_0                   │73GB      │2aa27df56a63 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q2_K                   │29GB      │b45d7449cc92 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q3_K_S                 │30GB      │2227129bfc8a │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q3_K_M                 │33GB      │fbd62f5dcf16 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q3_K_L                 │36GB      │b5bdb4982992 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q4_K_S                 │39GB      │1197a01c76e7 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q4_K_M                 │41GB      │bc70b8e5b4e5 │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q5_K_S                 │47GB      │eab38b764b5b │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q5_K_M                 │49GB      │b365053e53fc │Updated 2 months ago │                                                                                                                                                                                         
│   ├───llama2:70b-text-q6_K                   │57GB      │78e480a0450b │Updated 2 months ago │                                                                                                                                                                                         
│   └───llama2:70b-text-fp16                   │138GB     │dc03bc3a16cf │Updated 2 months ago │                                                                                                                                                                                         
├───mistral_container                          │0.00GB    │             │                     │The 7B model released by Mistral AI, updated to version 0.2.                                                                                                                             
├───mixtral_container                          │0.00GB    │             │                     │A high-quality Mixture of Experts (MoE) model with open weights by Mistral AI.                                                                                                           
├───llava_container                            │0.00GB    │             │                     │🌋 LLaVA is a novel end-to-end trained large multimodal model that combines a vision encoder and Vicuna for general-purpose visual and language understanding. Updated to version 1.6.    
├───neural-chat_container                      │0.00GB    │             │                     │A fine-tuned model based on Mistral with good coverage of domain and language.                                                                                                           
├───codellama_container                        │0.00GB    │             │                     │A large language model that can use text prompts to generate and discuss code.                                                                                                           
├───dolphin-mixtral_container                  │0.00GB    │             │                     │An uncensored, fine-tuned model based on the Mixtral mixture of experts model that excels at coding tasks. Created by Eric Hartford.                                                     
├───mistral-openorca_container                 │0.00GB    │             │                     │Mistral OpenOrca is a 7 billion parameter model, fine-tuned on top of the Mistral 7B model using the OpenOrca dataset.                                                                   
├───qwen_container                             │0.00GB    │             │                     │Qwen 1.5 is a series of large language models by Alibaba Cloud spanning from 0.5B to 72B parameters                                                                                      
├───llama2-uncensored_container                │0.00GB    │             │                     │Uncensored Llama 2 model by George Sung and Jarrad Hope.                                                                                                                                 
├───nous-hermes2_container                     │0.00GB    │             │                     │The powerful family of models by Nous Research that excels at scientific discussion and coding tasks.                                                                                    
├───phi_container                              │0.00GB    │             │                     │Phi-2: a 2.7B language model by Microsoft Research that demonstrates outstanding reasoning and language understanding capabilities.                                                      
├───deepseek-coder_container                   │0.00GB    │             │                     │DeepSeek Coder is a capable coding model trained on two trillion code and natural language tokens.                                                                                       
├───orca-mini_container                        │0.00GB    │             │                     │A general-purpose model ranging from 3 billion parameters to 70 billion, suitable for entry-level hardware.                                                                              
├───dolphin-mistral_container                  │0.00GB    │             │                     │The uncensored Dolphin model based on Mistral that excels at coding tasks. Updated to version 2.6.                                                                                       
├───wizard-vicuna-uncensored_container         │0.00GB    │             │                     │Wizard Vicuna Uncensored is a 7B, 13B, and 30B parameter model based on Llama 2 uncensored by Eric Hartford.                                                                             
├───vicuna_container                           │0.00GB    │             │                     │General use chat model based on Llama and Llama 2 with 2K to 16K context sizes.                                                                                                          
├───zephyr_container                           │0.00GB    │             │                     │Zephyr beta is a fine-tuned 7B version of mistral that was trained on on a mix of publicly available, synthetic datasets.                                                                
├───openhermes_container                       │0.00GB    │             │                     │OpenHermes 2.5 is a 7B model fine-tuned by Teknium on Mistral with fully open datasets.                                                                                                  
├───llama2-chinese_container                   │0.00GB    │             │                     │Llama 2 based model fine tuned to improve Chinese dialogue ability.                                                                                                                      
├───tinyllama_container                        │0.00GB    │             │                     │The TinyLlama project is an open endeavor to train a compact 1.1B Llama model on 3 trillion tokens.                                                                                      
├───wizardcoder_container                      │0.00GB    │             │                     │State-of-the-art code generation model                                                                                                                                                   
├───openchat_container                         │0.00GB    │             │                     │A family of open-source models trained on a wide variety of data, surpassing ChatGPT on various benchmarks. Updated to version 3.5-0106.                                                 
├───phind-codellama_container                  │0.00GB    │             │                     │Code generation model based on Code Llama.                                                                                                                                               
├───tinydolphin_container                      │0.00GB    │             │                     │An experimental 1.1B parameter model trained on the new Dolphin 2.8 dataset by Eric Hartford and based on TinyLlama.                                                                     
├───orca2_container                            │0.00GB    │             │                     │Orca 2 is built by Microsoft research, and are a fine-tuned version of Meta's Llama 2 models.  The model is designed to excel particularly in reasoning.                                 
├───falcon_container                           │0.00GB    │             │                     │A large language model built by the Technology Innovation Institute (TII) for use in summarization, text generation, and chat bots.                                                      
├───starcoder_container                        │0.00GB    │             │                     │StarCoder is a code generation model trained on 80+ programming languages.                                                                                                               
├───yi_container                               │0.00GB    │             │                     │A high-performing, bilingual language model.                                                                                                                                             
├───wizard-math_container                      │0.00GB    │             │                     │Model focused on math and logic problems                                                                                                                                                 
├───dolphin-phi_container                      │0.00GB    │             │                     │2.7B uncensored Dolphin model by Eric Hartford, based on the Phi language model by Microsoft Research.                                                                                   
├───nous-hermes_container                      │0.00GB    │             │                     │General use models based on Llama and Llama 2 from Nous Research.                                                                                                                        
├───nomic-embed-text_container                 │0.00GB    │             │                     │A high-performing open embedding model with a large token context window.                                                                                                                
├───starling-lm_container                      │0.00GB    │             │                     │Starling is a large language model trained by reinforcement learning from AI feedback focused on improving chatbot helpfulness.                                                          
├───starcoder2_container                       │0.00GB    │             │                     │StarCoder2 is the next generation of transparently trained open code LLMs that comes in three sizes: 3B, 7B and 15B parameters.                                                          
├───stable-code_container                      │0.00GB    │             │                     │Stable Code 3B is a model offering accurate and responsive code completion at a level on par with models such as CodeLLaMA 7B that are 2.5x larger.                                      
├───medllama2_container                        │0.00GB    │             │                     │Fine-tuned Llama 2 model to answer medical questions based on an open source medical dataset.                                                                                            
├───codeup_container                           │0.00GB    │             │                     │Great code generation model based on Llama2.                                                                                                                                             
├───bakllava_container                         │0.00GB    │             │                     │BakLLaVA is a multimodal model consisting of the Mistral 7B base model augmented with the LLaVA  architecture.                                                                           
├───wizardlm-uncensored_container              │0.00GB    │             │                     │Uncensored version of Wizard LM model                                                                                                                                                    
├───solar_container                            │0.00GB    │             │                     │A compact, yet powerful 10.7B large language model designed for single-turn conversation.                                                                                                
├───everythinglm_container                     │0.00GB    │             │                     │Uncensored Llama2 based model with support for a 16K context window.                                                                                                                     
├───sqlcoder_container                         │0.00GB    │             │                     │SQLCoder is a code completion model fined-tuned on StarCoder for SQL generation tasks                                                                                                    
├───stable-beluga_container                    │0.00GB    │             │                     │Llama 2 based model fine tuned on an Orca-style dataset. Originally called Free Willy.                                                                                                   
├───nous-hermes2-mixtral_container             │0.00GB    │             │                     │The Nous Hermes 2 model from Nous Research, now trained over Mixtral.                                                                                                                    
├───yarn-mistral_container                     │0.00GB    │             │                     │An extension of Mistral to support context windows of 64K or 128K.                                                                                                                       
├───samantha-mistral_container                 │0.00GB    │             │                     │A companion assistant trained in philosophy, psychology, and personal relationships. Based on Mistral.                                                                                   
├───meditron_container                         │0.00GB    │             │                     │Open-source medical large language model adapted from Llama 2 to the medical domain.                                                                                                     
├───stablelm2_container                        │0.00GB    │             │                     │Stable LM 2 1.6B is a state-of-the-art 1.6 billion parameter small language model trained on multilingual data in English, Spanish, German, Italian, French, Portuguese, and Dutch.      
├───stablelm-zephyr_container                  │0.00GB    │             │                     │A lightweight chat model allowing accurate, and responsive output without requiring high-end hardware.                                                                                   
├───magicoder_container                        │0.00GB    │             │                     │🎩 Magicoder is a family of 7B parameter models trained on 75K synthetic instruction data using OSS-Instruct, a novel approach to enlightening LLMs with open-source code snippets.       
├───yarn-llama2_container                      │0.00GB    │             │                     │An extension of Llama 2 that supports a context of up to 128k tokens.                                                                                                                    
├───wizard-vicuna_container                    │0.00GB    │             │                     │Wizard Vicuna is a 13B parameter model based on Llama 2 trained by MelodysDreamj.                                                                                                        
├───deepseek-llm_container                     │0.00GB    │             │                     │An advanced language model crafted with 2 trillion bilingual tokens.                                                                                                                     
├───llama-pro_container                        │0.00GB    │             │                     │An expansion of Llama 2 that specializes in integrating both general language understanding and domain-specific knowledge, particularly in programming and mathematics.                  
├───mistrallite_container                      │0.00GB    │             │                     │MistralLite is a fine-tuned model based on Mistral with enhanced capabilities of processing long contexts.                                                                               
├───codebooga_container                        │0.00GB    │             │                     │A high-performing code instruct model created by merging two existing code models.                                                                                                       
├───nexusraven_container                       │0.00GB    │             │                     │Nexus Raven is a 13B instruction tuned model for function calling tasks.                                                                                                                 
├───open-orca-platypus2_container              │0.00GB    │             │                     │Merge of the Open Orca OpenChat model and the Garage-bAInd Platypus 2 model. Designed for chat and code generation.                                                                      
├───goliath_container                          │0.00GB    │             │                     │A language model created by combining two fine-tuned Llama 2 70B models into one.                                                                                                        
├───notux_container                            │0.00GB    │             │                     │A top-performing mixture of experts model, fine-tuned with high-quality data.                                                                                                            
├───dolphincoder_container                     │0.00GB    │             │                     │An uncensored variant of the Dolphin model family that excels at coding, based on StarCoder2.                                                                                            
├───alfred_container                           │0.00GB    │             │                     │A robust conversational model designed to be used for both chat and instruct use cases.                                                                                                  
├───megadolphin_container                      │0.00GB    │             │                     │MegaDolphin-2.2-120b is a transformation of Dolphin-2.2-70b created by interleaving the model with itself.                                                                               
├───xwinlm_container                           │0.00GB    │             │                     │Conversational model based on Llama 2 that performs competitively on various benchmarks.                                                                                                 
├───wizardlm_container                         │0.00GB    │             │                     │General use 70 billion parameter model based on Llama 2.                                                                                                                                 
├───all-minilm_container                       │0.00GB    │             │                     │Embedding models on very large sentence level datasets.                                                                                                                                  
├───duckdb-nsql_container                      │0.00GB    │             │                     │7B parameter text-to-SQL model made by MotherDuck and Numbers Station.                                                                                                                   
└───notus_container                            │0.00GB    │             │                     │A 7B chat model fine-tuned with high-quality data and based on Zephyr.                                                                                                                   

```

**Comment by kkoreilly** [comment]
I will work on implementing the tree table widget as soon as I have time.

**Comment by ddkwork** [comment]
Very much looking forward to it, I'm going to implement a file explorer with it.
```go
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
│Name                                                                                  │Size   │Type        │ModTime             │Path                                                                          
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
│.                                                                                     │444575 │.           │0001-01-01 00:00:00 │                                                                              
├───Group1_container.json                                                              │4024   │----------  │2024-03-27 01:45:28 │Group1_container.json                                                         
├───Packed2_container.json                                                             │807    │----------  │2024-03-25 21:30:08 │Packed2_container.json                                                        
├───json2go                                                                            │57472  │json2go     │2024-03-27 22:14:07 │json2go                                                                       
│   ├───NodeType_enum_gen.go                                                           │1959   │----------  │2024-03-27 05:59:38 │json2go\NodeType_enum_gen.go                                                  
│   ├───ToGo.go                                                                        │3281   │----------  │2024-03-27 01:49:37 │json2go\ToGo.go                                                               
│   ├───config.json                                                                    │1464   │----------  │2024-01-26 17:02:18 │json2go\config.json                                                           
│   ├───game.json                                                                      │785    │----------  │2024-01-26 16:58:05 │json2go\game.json                                                             
│   ├───go.mod                                                                         │1543   │----------  │2024-03-27 21:21:39 │json2go\go.mod                                                                
│   ├───go.sum                                                                         │12568  │----------  │2024-03-27 21:21:39 │json2go\go.sum                                                                
│   ├───json2go                                                                        │8260   │json2go     │2024-03-27 17:30:49 │json2go\json2go                                                               
│   │   ├───json2ast.go                                                                │5710   │----------  │2024-03-27 17:30:49 │json2go\json2go\json2ast.go                                                   
│   │   ├───json2ast_test.go                                                           │1819   │----------  │2024-03-27 17:30:49 │json2go\json2go\json2ast_test.go                                              
│   │   └───util.go                                                                    │731    │----------  │2024-03-27 17:30:49 │json2go\json2go\util.go                                                       
│   ├───json2go.go                                                                     │4250   │----------  │2024-03-27 17:30:49 │json2go\json2go.go                                                            
│   ├───json2go_test.go                                                                │302    │----------  │2024-03-27 04:48:10 │json2go\json2go_test.go                                                       
│   ├───log.log                                                                        │175    │----------  │2024-03-27 22:15:04 │json2go\log.log                                                               
│   ├───product-info.json                                                              │13085  │----------  │2024-01-26 16:58:05 │json2go\product-info.json                                                     
│   ├───std.go                                                                         │9629   │----------  │2024-03-27 22:14:07 │json2go\std.go                                                                
│   └───std_test.go                                                                    │171    │----------  │2024-03-27 21:55:30 │json2go\std_test.go                                                           
├───log.log                                                                            │0      │----------  │2024-03-27 22:26:02 │log.log                                                                       
├───protobuf                                                                           │363954 │protobuf    │2024-03-27 17:30:49 │protobuf                                                                      
│   ├───.git                                                                           │251405 │.git        │2024-03-27 21:22:32 │protobuf\.git                                                                 
│   │   ├───COMMIT_EDITMSG                                                             │14     │----------  │2024-03-27 21:22:32 │protobuf\.git\COMMIT_EDITMSG                                                  
│   │   ├───FETCH_HEAD                                                                 │906    │----------  │2024-03-27 16:17:13 │protobuf\.git\FETCH_HEAD                                                      
│   │   ├───HEAD                                                                       │23     │----------  │2024-03-23 09:42:36 │protobuf\.git\HEAD                                                            
│   │   ├───config                                                                     │299    │----------  │2024-03-23 09:42:36 │protobuf\.git\config                                                          
│   │   ├───description                                                                │73     │----------  │2024-03-23 09:42:33 │protobuf\.git\description                                                     
│   │   ├───hooks                                                                      │23513  │hooks       │2024-03-23 09:42:33 │protobuf\.git\hooks                                                           
│   │   │   ├───applypatch-msg.sample                                                  │478    │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\applypatch-msg.sample                                     
│   │   │   ├───commit-msg.sample                                                      │896    │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\commit-msg.sample                                         
│   │   │   ├───fsmonitor-watchman.sample                                              │4726   │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\fsmonitor-watchman.sample                                 
│   │   │   ├───post-update.sample                                                     │189    │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\post-update.sample                                        
│   │   │   ├───pre-applypatch.sample                                                  │424    │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\pre-applypatch.sample                                     
│   │   │   ├───pre-commit.sample                                                      │1643   │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\pre-commit.sample                                         
│   │   │   ├───pre-merge-commit.sample                                                │416    │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\pre-merge-commit.sample                                   
│   │   │   ├───pre-push.sample                                                        │1374   │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\pre-push.sample                                           
│   │   │   ├───pre-rebase.sample                                                      │4898   │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\pre-rebase.sample                                         
│   │   │   ├───pre-receive.sample                                                     │544    │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\pre-receive.sample                                        
│   │   │   ├───prepare-commit-msg.sample                                              │1492   │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\prepare-commit-msg.sample                                 
│   │   │   ├───push-to-checkout.sample                                                │2783   │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\push-to-checkout.sample                                   
│   │   │   └───update.sample                                                          │3650   │----------  │2024-03-23 09:42:33 │protobuf\.git\hooks\update.sample                                             
│   │   ├───index                                                                      │2338   │----------  │2024-03-27 21:22:32 │protobuf\.git\index                                                           
│   │   ├───info                                                                       │240    │info        │2024-03-23 09:42:33 │protobuf\.git\info                                                            
│   │   │   └───exclude                                                                │240    │----------  │2024-03-23 09:42:33 │protobuf\.git\info\exclude                                                    
│   │   ├───logs                                                                       │1444   │logs        │2024-03-23 09:42:36 │protobuf\.git\logs                                                            
│   │   │   ├───HEAD                                                                   │635    │----------  │2024-03-27 21:22:32 │protobuf\.git\logs\HEAD                                                       
│   │   │   └───refs                                                                   │809    │refs        │2024-03-23 09:42:36 │protobuf\.git\logs\refs                                                       
│   │   │   │   ├───heads                                                              │635    │heads       │2024-03-23 09:42:36 │protobuf\.git\logs\refs\heads                                                 
│   │   │   │   │   └───master                                                         │635    │----------  │2024-03-27 21:22:32 │protobuf\.git\logs\refs\heads\master                                          
│   │   │   │   └───remotes                                                            │174    │remotes     │2024-03-23 09:42:36 │protobuf\.git\logs\refs\remotes                                               
│   │   │   │   │   └───origin                                                         │174    │origin      │2024-03-23 09:42:36 │protobuf\.git\logs\refs\remotes\origin                                        
│   │   │   │   │   │   └───HEAD                                                       │174    │----------  │2024-03-23 09:42:36 │protobuf\.git\logs\refs\remotes\origin\HEAD                                   
│   │   ├───objects                                                                    │220998 │objects     │2024-03-27 17:47:59 │protobuf\.git\objects                                                         
│   │   │   ├───15                                                                     │303    │15          │2024-03-27 17:41:42 │protobuf\.git\objects\15                                                      
│   │   │   │   └───eda15801fbb3229d5ef32903b0394350149505                             │303    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\15\eda15801fbb3229d5ef32903b0394350149505               
│   │   │   ├───2c                                                                     │154    │2c          │2024-03-27 21:22:32 │protobuf\.git\objects\2c                                                      
│   │   │   │   └───d803bd2e2461aff62b9873ffe920e1838323ed                             │154    │----------  │2024-03-27 21:22:32 │protobuf\.git\objects\2c\d803bd2e2461aff62b9873ffe920e1838323ed               
│   │   │   ├───32                                                                     │295    │32          │2024-03-27 17:41:42 │protobuf\.git\objects\32                                                      
│   │   │   │   └───801ef8080e37e5681cef9c1626bb9e0bff2c94                             │295    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\32\801ef8080e37e5681cef9c1626bb9e0bff2c94               
│   │   │   ├───3b                                                                     │2617   │3b          │2024-03-27 17:41:42 │protobuf\.git\objects\3b                                                      
│   │   │   │   └───580b1b34ca0b8ba3fc4bcb37247305b82b7779                             │2617   │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\3b\580b1b34ca0b8ba3fc4bcb37247305b82b7779               
│   │   │   ├───3d                                                                     │453    │3d          │2024-03-27 17:41:42 │protobuf\.git\objects\3d                                                      
│   │   │   │   └───64b4e403fa454cc7d354359b293ee136ea1908                             │453    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\3d\64b4e403fa454cc7d354359b293ee136ea1908               
│   │   │   ├───43                                                                     │904    │43          │2024-03-27 17:47:59 │protobuf\.git\objects\43                                                      
│   │   │   │   └───e7e0d696c39bb3d7517e41c983e307efaecef7                             │904    │----------  │2024-03-27 17:47:59 │protobuf\.git\objects\43\e7e0d696c39bb3d7517e41c983e307efaecef7               
│   │   │   ├───46                                                                     │325    │46          │2024-03-27 17:41:42 │protobuf\.git\objects\46                                                      
│   │   │   │   └───a9b21d04d57e589fce4417de11ed1b950a1ba2                             │325    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\46\a9b21d04d57e589fce4417de11ed1b950a1ba2               
│   │   │   ├───55                                                                     │162    │55          │2024-03-27 17:41:42 │protobuf\.git\objects\55                                                      
│   │   │   │   └───eddbc2d851e6778075d5256104a16b10830c63                             │162    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\55\eddbc2d851e6778075d5256104a16b10830c63               
│   │   │   ├───5a                                                                     │526    │5a          │2024-03-27 17:41:42 │protobuf\.git\objects\5a                                                      
│   │   │   │   └───0c585178326e6cd8b991d5a197ca0eea152088                             │526    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\5a\0c585178326e6cd8b991d5a197ca0eea152088               
│   │   │   ├───5c                                                                     │6310   │5c          │2024-03-27 21:22:32 │protobuf\.git\objects\5c                                                      
│   │   │   │   └───55c3fc36fa71da6057172b164657b6c9ded0cf                             │6310   │----------  │2024-03-27 21:22:32 │protobuf\.git\objects\5c\55c3fc36fa71da6057172b164657b6c9ded0cf               
│   │   │   ├───66                                                                     │5213   │66          │2024-03-27 17:41:42 │protobuf\.git\objects\66                                                      
│   │   │   │   └───7bcf5069c77710482bb507c53a843fcdf3be3f                             │5213   │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\66\7bcf5069c77710482bb507c53a843fcdf3be3f               
│   │   │   ├───68                                                                     │263    │68          │2024-03-27 17:41:42 │protobuf\.git\objects\68                                                      
│   │   │   │   └───944efd6530bf766cf5bb23c9d81aa5706e1e13                             │263    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\68\944efd6530bf766cf5bb23c9d81aa5706e1e13               
│   │   │   ├───73                                                                     │2362   │73          │2024-03-27 17:41:42 │protobuf\.git\objects\73                                                      
│   │   │   │   └───61ead7a4461a20b1f83e2a9bf98792ba21bdf7                             │2362   │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\73\61ead7a4461a20b1f83e2a9bf98792ba21bdf7               
│   │   │   ├───7a                                                                     │904    │7a          │2024-03-27 17:41:42 │protobuf\.git\objects\7a                                                      
│   │   │   │   └───dbe34a9b69ec2beb4a89990d876e305f872d19                             │904    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\7a\dbe34a9b69ec2beb4a89990d876e305f872d19               
│   │   │   ├───8b                                                                     │759    │8b          │2024-03-27 21:22:32 │protobuf\.git\objects\8b                                                      
│   │   │   │   └───60c9a998fad8ecf9e7a4962ecc742e715bcd13                             │759    │----------  │2024-03-27 21:22:32 │protobuf\.git\objects\8b\60c9a998fad8ecf9e7a4962ecc742e715bcd13               
│   │   │   ├───8f                                                                     │1311   │8f          │2024-03-27 17:41:42 │protobuf\.git\objects\8f                                                      
│   │   │   │   └───0d82a04771371e2975270d69e44421aae75d7c                             │1311   │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\8f\0d82a04771371e2975270d69e44421aae75d7c               
│   │   │   ├───9b                                                                     │297    │9b          │2024-03-27 17:41:42 │protobuf\.git\objects\9b                                                      
│   │   │   │   └───34073193cc4aef69a690015d5649eb811d6cd3                             │297    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\9b\34073193cc4aef69a690015d5649eb811d6cd3               
│   │   │   ├───9c                                                                     │764    │9c          │2024-03-27 17:41:42 │protobuf\.git\objects\9c                                                      
│   │   │   │   └───b3f63d2d93b0ad02a560c278261feb42de63c4                             │764    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\9c\b3f63d2d93b0ad02a560c278261feb42de63c4               
│   │   │   ├───af                                                                     │1003   │af          │2024-03-27 17:41:42 │protobuf\.git\objects\af                                                      
│   │   │   │   └───39185f5b53ad92d2ab4cb20dc1368e19cbc0f5                             │1003   │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\af\39185f5b53ad92d2ab4cb20dc1368e19cbc0f5               
│   │   │   ├───b9                                                                     │397    │b9          │2024-03-27 17:47:59 │protobuf\.git\objects\b9                                                      
│   │   │   │   └───d04de0ba26c7ea2a643221762281919a8f4f97                             │397    │----------  │2024-03-27 17:47:59 │protobuf\.git\objects\b9\d04de0ba26c7ea2a643221762281919a8f4f97               
│   │   │   ├───d5                                                                     │2114   │d5          │2024-03-27 17:47:59 │protobuf\.git\objects\d5                                                      
│   │   │   │   └───6bc3ff7ecae2b61b090a51ab184584ca7e7ce2                             │2114   │----------  │2024-03-27 17:47:59 │protobuf\.git\objects\d5\6bc3ff7ecae2b61b090a51ab184584ca7e7ce2               
│   │   │   ├───e4                                                                     │907    │e4          │2024-03-27 21:22:32 │protobuf\.git\objects\e4                                                      
│   │   │   │   └───d01ab77c6e470345d750d5150e1cbdadccfd6c                             │907    │----------  │2024-03-27 21:22:32 │protobuf\.git\objects\e4\d01ab77c6e470345d750d5150e1cbdadccfd6c               
│   │   │   ├───e6                                                                     │457    │e6          │2024-03-27 17:47:59 │protobuf\.git\objects\e6                                                      
│   │   │   │   ├───9b7251b3d0e4ab6e0481b7589a3f13c00214e3                             │293    │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\e6\9b7251b3d0e4ab6e0481b7589a3f13c00214e3               
│   │   │   │   └───a4512b63df8ee3ae934baa2aae099c74f20f53                             │164    │----------  │2024-03-27 17:47:59 │protobuf\.git\objects\e6\a4512b63df8ee3ae934baa2aae099c74f20f53               
│   │   │   ├───f6                                                                     │2440   │f6          │2024-03-27 17:41:42 │protobuf\.git\objects\f6                                                      
│   │   │   │   └───9ba2fba8f629ebb560b85e602e163349a488d9                             │2440   │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\f6\9ba2fba8f629ebb560b85e602e163349a488d9               
│   │   │   ├───f9                                                                     │3756   │f9          │2024-03-27 17:41:42 │protobuf\.git\objects\f9                                                      
│   │   │   │   └───3974db2d212e6e6b9355ca80ac1e0e95c1deac                             │3756   │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\f9\3974db2d212e6e6b9355ca80ac1e0e95c1deac               
│   │   │   ├───fc                                                                     │1889   │fc          │2024-03-27 17:41:42 │protobuf\.git\objects\fc                                                      
│   │   │   │   └───1cd8da5f8718fbedb2dd1868a52c0309749e2d                             │1889   │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\fc\1cd8da5f8718fbedb2dd1868a52c0309749e2d               
│   │   │   ├───info                                                                   │0      │info        │2024-03-23 09:42:33 │protobuf\.git\objects\info                                                    
│   │   │   └───pack                                                                   │184113 │pack        │2024-03-23 09:42:36 │protobuf\.git\objects\pack                                                    
│   │   │   │   ├───pack-d46b988338a0cfad738dc29d7a8eebb04aab772e.idx                  │15940  │----------  │2024-03-23 09:42:36 │protobuf\.git\objects\pack\pack-d46b988338a0cfad738dc29d7a8eebb04aab772e.idx  
│   │   │   │   └───pack-d46b988338a0cfad738dc29d7a8eebb04aab772e.pack                 │168173 │----------  │2024-03-27 17:41:42 │protobuf\.git\objects\pack\pack-d46b988338a0cfad738dc29d7a8eebb04aab772e.pack 
│   │   ├───packed-refs                                                                │1484   │----------  │2024-03-23 09:42:36 │protobuf\.git\packed-refs                                                     
│   │   └───refs                                                                       │73     │refs        │2024-03-23 09:42:36 │protobuf\.git\refs                                                            
│   │   │   ├───heads                                                                  │41     │heads       │2024-03-27 21:22:32 │protobuf\.git\refs\heads                                                      
│   │   │   │   └───master                                                             │41     │----------  │2024-03-27 21:22:32 │protobuf\.git\refs\heads\master                                               
│   │   │   ├───remotes                                                                │32     │remotes     │2024-03-23 09:42:36 │protobuf\.git\refs\remotes                                                    
│   │   │   │   └───origin                                                             │32     │origin      │2024-03-23 09:42:36 │protobuf\.git\refs\remotes\origin                                             
│   │   │   │   │   └───HEAD                                                           │32     │----------  │2024-03-23 09:42:36 │protobuf\.git\refs\remotes\origin\HEAD                                        
│   │   │   └───tags                                                                   │0      │tags        │2024-03-23 09:42:33 │protobuf\.git\refs\tags                                                       
│   ├───.gitignore                                                                     │30     │----------  │2024-03-23 09:42:36 │protobuf\.gitignore                                                           
│   ├───.travis.yml                                                                    │221    │----------  │2024-03-23 09:42:36 │protobuf\.travis.yml                                                          
│   ├───COPYING                                                                        │18431  │----------  │2024-03-23 09:42:36 │protobuf\COPYING                                                              
│   ├───README.md                                                                      │7165   │----------  │2024-03-23 09:42:36 │protobuf\README.md                                                            
│   ├───array_test.go                                                                  │1063   │----------  │2024-03-27 17:30:49 │protobuf\array_test.go                                                        
│   ├───constructor_test.go                                                            │319    │----------  │2024-03-27 17:30:49 │protobuf\constructor_test.go                                                  
│   ├───crashers_test.go                                                               │1021   │----------  │2024-03-27 17:30:49 │protobuf\crashers_test.go                                                     
│   ├───decode.go                                                                      │14154  │----------  │2024-03-27 17:30:49 │protobuf\decode.go                                                            
│   ├───doc.go                                                                         │5763   │----------  │2024-03-27 17:30:49 │protobuf\doc.go                                                               
│   ├───encode.go                                                                      │11602  │----------  │2024-03-27 17:30:49 │protobuf\encode.go                                                            
│   ├───encoding_test.go                                                               │7636   │----------  │2024-03-27 17:30:49 │protobuf\encoding_test.go                                                     
│   ├───field.go                                                                       │2115   │----------  │2024-03-27 17:30:49 │protobuf\field.go                                                             
│   ├───field_test.go                                                                  │1311   │----------  │2024-03-27 17:30:49 │protobuf\field_test.go                                                        
│   ├───fuzz.go                                                                        │841    │----------  │2024-03-27 17:30:49 │protobuf\fuzz.go                                                              
│   ├───generate.go                                                                    │6645   │----------  │2024-03-27 17:30:49 │protobuf\generate.go                                                          
│   ├───generate_test.go                                                               │3628   │----------  │2024-03-27 17:30:49 │protobuf\generate_test.go                                                     
│   ├───go.mod                                                                         │1938   │----------  │2024-03-27 21:21:53 │protobuf\go.mod                                                               
│   ├───go.sum                                                                         │12310  │----------  │2024-03-27 21:21:53 │protobuf\go.sum                                                               
│   ├───interface.go                                                                   │1390   │----------  │2024-03-27 17:30:49 │protobuf\interface.go                                                         
│   ├───interface_test.go                                                              │905    │----------  │2024-03-27 17:30:49 │protobuf\interface_test.go                                                    
│   ├───map_test.go                                                                    │3024   │----------  │2024-03-27 17:30:49 │protobuf\map_test.go                                                          
│   ├───marshal_test.go                                                                │616    │----------  │2024-03-27 17:30:49 │protobuf\marshal_test.go                                                      
│   ├───optionalBytes_test.go                                                          │423    │----------  │2024-03-27 17:30:49 │protobuf\optionalBytes_test.go                                                
│   ├───person_test.go                                                                 │1501   │----------  │2024-03-27 17:30:49 │protobuf\person_test.go                                                       
│   ├───privateFields_test.go                                                          │544    │----------  │2024-03-27 17:30:49 │protobuf\privateFields_test.go                                                
│   ├───protobuf_test.go                                                               │6624   │----------  │2024-03-27 17:30:49 │protobuf\protobuf_test.go                                                     
│   ├───test1_test.go                                                                  │444    │----------  │2024-03-27 17:30:49 │protobuf\test1_test.go                                                        
│   ├───test2_test.go                                                                  │428    │----------  │2024-03-27 17:30:49 │protobuf\test2_test.go                                                        
│   └───test3_test.go                                                                  │457    │----------  │2024-03-27 17:30:49 │protobuf\test3_test.go                                                        
├───struct2table.go                                                                    │9434   │----------  │2024-03-27 17:30:49 │struct2table.go                                                               
├───struct2table_test.go                                                               │2447   │----------  │2024-03-27 17:30:49 │struct2table_test.go                                                          
├───structBytes                                                                        │3975   │structBytes │2024-03-23 23:26:03 │structBytes                                                                   
│   ├───BytesToStruct.go                                                               │1361   │----------  │2024-02-22 21:00:54 │structBytes\BytesToStruct.go                                                  
│   ├───StructToBytes.go                                                               │1161   │----------  │2024-02-22 21:00:54 │structBytes\StructToBytes.go                                                  
│   ├───encoding_test.go                                                               │623    │----------  │2024-02-19 18:22:37 │structBytes\encoding_test.go                                                  
│   └───interface.go                                                                   │830    │----------  │2024-03-23 09:46:05 │structBytes\interface.go                                                      
└───tree_kind.go                                                                       │2462   │----------  │2024-03-27 17:30:49 │tree_kind.go                                                                  
```

**Comment by ddkwork** [comment]
I've implemented various tree tables with unison, so to save you, I'll close this topic, and also the other topics about the code editor can be closed, since I've implemented syntax coloring with unison.

![Screenshot_20240428_010453_com.realvnc.viewer.android.jpg](https://github.com/cogentcore/core/assets/19886504/4e6299b6-e2ac-4fd7-82b3-450487277d41)

![Screenshot_20240428_010559_com.realvnc.viewer.android.jpg](https://github.com/cogentcore/core/assets/19886504/ee6e4848-c4fa-4750-a127-6d1414221b74)

![Screenshot_20240428_010752_com.realvnc.viewer.android.jpg](https://github.com/cogentcore/core/assets/19886504/18fbf85f-c5d9-4e45-a7af-271481106abc)

**Comment by kkoreilly** [comment]
It is great that you have implemented tree tables in unison and fine that you do not want to use Cogent Core for them, but that does not mean that we do not want to implement tree tables eventually, so this issue is still relevant and will stay open.

**Comment by ddkwork** [comment]
I also maked another one use gio

https://github.com/ddkwork/ux

**Comment by kkoreilly** [comment]
@ddkwork Thanks for the update; that looks cool. We will implement a tree table widget in Cogent Core later once we have the time.
